exercises/07-normalizr/src/modules/users/users-saga.ts

```diff
@@ -0,0 +1,55 @@
+import { all, call, fork, put, takeEvery } from 'redux-saga/effects'
+import { AxiosResponse } from 'axios'
+import { normalize } from 'normalizr'
+
+import * as Schema from 'modules/entities/entities-schema'
+import * as EntitiesTypes from 'modules/entities/entities-types'
+
+import { usersActions } from 'modules/users/users-slice'
+import * as UsersEffects from 'modules/users/users-effects'
+import { User } from './user-types'
+
+const logError = (error: unknown) => {
+  if (error instanceof Error) {
+    console.log(error.message)
+  } else {
+    console.log(error)
+  }
+}
+
+function* getUsers() {
+  try {
+    const { data: users }: AxiosResponse<User[]> = yield call(
+      UsersEffects.getUsers
+    )
+
+    const normalizedUsers = normalize<
+      User[],
+      EntitiesTypes.UserEntities,
+      EntitiesTypes.UserIds
+    >(users, Schema.users)
+
+    yield put(usersActions.usersLoaded(normalizedUsers))
```

Comment:
I would split this into 2 actions
- save `entities` ("generic" action, something like `entitiesActions.storeEntities`)
- save `ids` (`usersActions.usersLoaded`)

Reply:
I've given this some thought when I was implementing it. I've decided to use a single action for both these updates to the state, because this allowed to make the action dispatching more atomic. As only one action is dispatched, there's less possibility for the error. The presence of two actions creates two additional possible error scenarios:
  - one of the actions may be forgotten 
  - the order in which the actions are dispatched may be incorrect: when the `id`s are dispatched first, the `UserList` component will be re-rendered, but the `entities` won't be yet in the store  
 

Additionally I've read the Redux styleguide, specifically: 
  - [Allow Many Reducers to Respond to the Same Action](https://redux.js.org/style-guide/style-guide#allow-many-reducers-to-respond-to-the-same-action)
  - [Avoid Dispatching Many Actions Sequentially](https://redux.js.org/style-guide/style-guide#avoid-dispatching-many-actions-sequentially),

and figured it would be the best thing to do.

Reply:
> to make the action dispatching more atomic

Doing this in general: :+1: And I agree with both principles from the Redux styleguide in general but everything has its cost and is only about a trade-off. :slightly_smiling_face: 

> As only one action is dispatched, there's less possibility for the error.

Really? I don't think it's that clear. :slightly_smiling_face: Let me explain below.

> one of the actions may be forgotten

In the same way, one can forget to extend `extraReducers` of `entitiesSlice`.

> the order in which the actions are dispatched may be incorrect: when the ids are dispatched first, the UserList component will be re-rendered, but the entities won't be yet in the store

Yes, but this can be also solved (and it actually solves also the first problem). In the original solution, we had a saga that:
- normalizes data (calls `normalize`),
- stores entities (dispatches something like `entitiesActions.storeEntities`),
- and returns ids.

So instead of:
- `yield call` API
- `normalize`
- `yield put` output of normalize

We could have:
- `yield call` API
- `yield call` the new saga `normalizeAndStore`
- `yield put` ids

Quite similar but you don't have to extend `extraReducers` of `entitiesSlice` for every API endpoint or related Redux action.

The downside of the old approach is an additional re-render. In the old training, we have used `connect` which automatically memoizes the components (no additional re-render then). We don't have this advantage with hooks now. We would have to memoize the components explicitly with `React.memo`. But... we don't care about re-renders that much in this app. If we will update any data anywhere in the state (`title` for example), the whole app and potentially even the user list (if it's rendered while updating `title`) would be re-rendered. :slightly_smiling_face: Do we really care about this in this app? I don't think so.

I'm not strongly against your solution but in a real application, I wouldn't like to end up with something like:

```ts
const entitiesSlice = createSlice({
  name: 'entities',
  initialState: initState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(usersActions.usersLoaded, updateEntities)
    builder.addCase(usersActions.skillsLoaded, updateEntities)
    builder.addCase(usersActions.fooLoaded, updateEntities)
    builder.addCase(townActions.townsLoaded, updateEntities)
    builder.addCase(townActions.barLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
    builder.addCase(bazActions.abcdLoaded, updateEntities)
  },
})
```

I would leave it up to @novellizator, he has the final word I guess. :slightly_smiling_face:

Reply:
@branoholy Thanks for the explanation, I see you point :slightly_smiling_face: